### ISSUE TITLE
Refactor training via the Hugging Face Trainer - MultiGPU, loss logging, bf16, etc.

### ISSUE THREAD

**Issue by tomaarsen** [UKPLab/sentence-transformers#2436]
Hello!

## Context
This issue describes a refactor that I am planning for a v3 release of Sentence Transformers.
See #1733 for a related PR.

## Details
Sentence Transformers currently relies on a `fit` method with a custom training loop. Despite working very well, user needs have outgrown that method. I believe that the `transformers` Trainer, as proposed in #1733 by @matthewfranglen & #1446 by @atreyasha would be a very future-proof approach of improving the training in Sentence Transformers.

Ideally, I would like to introduce a new Trainer that implements a superclass of all behaviour supported by the current `fit` method. Then, I can fully refactor the `fit` method to use the new Trainer below the hood. This should prevent any backwards compatibility issues & allow training in the same way as before, although a new way (via the Trainer itself) will now also be supported. Training via the Trainer should allow for more opportunities to tune the training arguments.

## Pro's
* Just like #2435, this feature request should allow for better future-proofing by adopting powerful functionality from `transformers` whose interface they would not quickly change.
* Sentence Transformers gains all benefits of the `transformers` Trainer. Some notable ones:
  * MultiGPU support.
  * Loss logging via W&B or Tensorboard & the terminal.
  * bf16 training.

## Cons
* The `Trainer` is not explicitly created for embedding tasks, so we have to override certain "behind-the-scenes" code, i.e. code that could be changed from version to version & break the Sentence Transformers training.
* The `Trainer` likes to work with `Dataset` instances, which may require sunsetting the `InputExample` class in favor of preprocessing data with `datasets`.
* It may be challenging to avoid breaking changes, e.g. training with both positive pairs & text-positive-negative triplets might be possible currently (Not 100% confident), and we'd like to keep that behaviour possible if it currently is, but that will be difficult due to `Datasets` and their fixed columns.
* It may be challenging to build the Trainer into a complete superset of all `fit` functionality, e.g. `fit` allows multi-task learning with multiple datasets and loss functions.

cc @bwanglzu @ir2718 @johneckberg @aamir-s18 as I know you're interested in my TODO list. I'll lead this development - I've already started on a local branch, although I'm open to comments once I make the draft PR.

- Tom Aarsen

**Comment by tomaarsen** [comment]
Resolved via #2449 & the v3.0 release.

- Tom Aarsen
